consensus/istanbul/backend/announce.go

```diff
@@ -17,405 +17,671 @@
+	sb.cachedAnnounceMsgsMu.RLock()
+	defer sb.cachedAnnounceMsgsMu.RUnlock()
+	// TODO:  Add support for the AnnounceMsg to contain multiple announce messages within it's payload
+	for _, valAddress := range valAddresses {
+		if cachedAnnounceMsgEntry, ok := sb.cachedAnnounceMsgs[valAddress]; ok {
+			logger.Trace("Sending announce msg", "peer", peer, "valAddress", valAddress)
+			go peer.Send(istanbulAnnounceMsg, cachedAnnounceMsgEntry.MsgPayload)
+		}
+	}
 
-	// Send out an announce message when this thread starts
-	go sb.sendIstAnnounce()
+	return nil
+}
 
-	// Set the initial states
-	announceThreadState := HighFreqBeforeFirstPeerState
-	currentTickerDuration := HighFreqTickerDuration
-	ticker := time.NewTicker(currentTickerDuration)
-	numSentMsgsInHighFreqAfterFirstPeerState := 0
+// This function will generate the lastest announce msg from this node (as opposed to retrieving from the announceMsgCache) and then broadcast it to it's peers,
+// which should then gossip the announce msg message throughout the p2p network, since this announce msg's timestamp should be the latest among all of this
+// validator's previous announce msgs.
+func (sb *Backend) gossipAnnounce() error {
+	logger := sb.logger.New("func", "gossipAnnounce")
+	logger.Trace("gossipAnnounce called")
+	istMsg, err := sb.generateAnnounce()
+	if err != nil {
+		return err
+	}
 
-	for {
-		select {
-		case <-sb.newEpochCh:
-			go sb.sendIstAnnounce()
+	if istMsg == nil {
+		return nil
+	}
 
-		case <-ticker.C:
-			switch announceThreadState {
-			case HighFreqBeforeFirstPeerState:
-				{
-					if len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) > 0 {
-						announceThreadState = HighFreqAfterFirstPeerState
-					}
-				}
+	// Sign the announce message
+	if err := istMsg.Sign(sb.Sign); err != nil {
+		logger.Error("Error in signing an Istanbul Announce Message", "AnnounceMsg", istMsg.String(), "err", err)
+		return err
+	}
 
-			case HighFreqAfterFirstPeerState:
-				{
-					if numSentMsgsInHighFreqAfterFirstPeerState >= 10 {
-						announceThreadState = LowFreqState
-					}
-					numSentMsgsInHighFreqAfterFirstPeerState += 1
-				}
+	// Convert to payload
+	payload, err := istMsg.Payload()
+	if err != nil {
+		logger.Error("Error in converting Istanbul Announce Message to payload", "AnnounceMsg", istMsg.String(), "err", err)
+		return err
+	}
 
-			case LowFreqState:
-				{
-					if currentTickerDuration != LowFreqTickerDuration {
-						// Reset the ticker
-						currentTickerDuration = LowFreqTickerDuration
-						ticker.Stop()
-						ticker = time.NewTicker(currentTickerDuration)
-					}
-				}
-			}
+	// Add the generated announce message to this node's cache
+	var announcePayload validatorEncryptedEnodes
+	rlp.DecodeBytes(istMsg.Msg, &announcePayload)
+	sb.cachedAnnounceMsgs[announcePayload.ValAddress] = &announceMsgCachedEntry{MsgVersion: announcePayload.Version,
+		MsgPayload: payload}
 
-			go sb.sendIstAnnounce()
+	sb.Multicast(nil, payload, istanbulAnnounceMsg)
```

Comment:
Even though we know, based on the current implementation, `sb.Multicast` will return a `nil` error, we should still handle it. One reason for this is that if someone updates it to return non-nil in the future they likely won't think to update the call sites. (This comment applied to all the call sites for `sb.Multicast` in this file)